aten/src/ATen/core/typeid.h

```diff
@@ -81,50 +90,144 @@ AT_DEFINE_HASH_FOR_IDWRAPPER(caffe2::TypeIdentifier)
 
-AT_CORE_API std::mutex& gTypeRegistrationMutex();
+/**
+ * Placement new function for the type.
+ */
+template <typename T>
+inline void _Ctor(void* ptr, size_t n) {
+  T* typed_ptr = static_cast<T*>(ptr);
+  for (size_t i = 0; i < n; ++i) {
+    new (typed_ptr + i) T;
+  }
+}
 
 template <typename T>
-struct TypeNameRegisterer {
-  TypeNameRegisterer(TypeIdentifier id, const std::string& literal_name) {
-    std::lock_guard<std::mutex> guard(gTypeRegistrationMutex());
-#ifdef __GXX_RTTI
-    (void)literal_name;
-
-    std::string name = at::demangle(typeid(T).name());
-    // If we are in RTTI mode, we will also use this opportunity to do sanity
-    // check if there are duplicated ids registered for the same type. This
-    // usually happens when one does not do RTLD_GLOBAL, which is often the
-    // case in Python. The way we do the check is to make sure that there are
-    // no duplicated names registered - this could be done by checking the
-    // uniqueness of names.
-    if (gRegisteredTypeNames().count(name)) {
-      AT_ERROR("typeid.h: Type name ", name, " was registered twice.  "
-               "This should not happen.  Things to check:\n"
-               "1. Did you add a new CAFFE_KNOWN_TYPE?  If so, check that "
-               "it is not duplicated with an existing CAFFE_KNOWN_TYPE.\n"
-               "2. Did you build and install PyTorch and Caffe2 separately? "
-               "For example, this would be the case if you ran scripts/onnx/install.sh or "
-               "scripts/onnx/install-develop.sh prior to Aug 12, 2018 "
-               "(commit 1756daaa7530d).  If so, rebuild using the environment variable "
-               " FULL_CAFFE2=1 (if you build latest master, the ONNX scripts are "
-               "updated to do this for you.) "
-               "For more context, see https://github.com/pytorch/pytorch/issues/10460");
-    }
-    gRegisteredTypeNames().insert(name);
-    gTypeNames()[id] = name;
-#else // __GXX_RTTI
-    if (literal_name.empty()) {
-      gTypeNames()[id] = "(RTTI disabled, cannot show name)";
-    } else {
-      gTypeNames()[id] = literal_name;
-    }
-#endif // __GXX_RTTI
+inline void _CtorNotDefault(void* /*ptr*/, size_t /*n*/) {
+  _ThrowRuntimeTypeLogicError(
+      "Type " + std::string(at::demangle_type<T>()) +
+      " is not default-constructible.");
+}
+
+template <
+    typename T,
+    typename std::enable_if<std::is_default_constructible<T>::value>::type* =
+        nullptr>
+inline TypeMetaData::PlacementNew* _PickCtor() {
+  return &_Ctor<T>;
+}
+
+template <
+    typename T,
+    typename std::enable_if<!std::is_default_constructible<T>::value>::type* =
+        nullptr>
+inline TypeMetaData::PlacementNew* _PickCtor() {
+  return &_CtorNotDefault<T>;
+}
+
+/**
+ * Typed copy function for classes.
+ */
+template <typename T>
+inline void _Copy(const void* src, void* dst, size_t n) {
+  const T* typed_src = static_cast<const T*>(src);
```

Comment:
I guess these are guaranteed not to overlap?

Reply:
This is not changed in this PR and worked before.

Reply:
The code is not correct if the inputs overlap. Worth fixing this, now that we've noticed.

Reply:
(Obviously, that can be in another PR.)